src/stackmap_checker/trace.c

```diff
@@ -0,0 +1,161 @@
+    // The index of the record which corresponds to the patchpoint which
+    // called __guard_failure.
+    int opt_rec_idx = stmap_get_map_record(sm, sm_id);
+
+    if (unopt_rec_idx == -1 || opt_rec_idx == -1) {
+        errx(1, "Stack map record not found. Exiting.\n");
+    }
+
+    stack_map_record_t unopt_rec = sm->stk_map_records[unopt_rec_idx];
+    stack_map_record_t opt_rec = sm->stk_map_records[opt_rec_idx];
+
+    // Save the direct (stack) locations. These need to be restored later,
+    // because the process of restoring them might cause other values on the
+    // stack of `trace` to be overwritten.
+    uint64_t *direct_locations =
+        (uint64_t *)malloc(sizeof(uint64_t) * unopt_rec.num_locations);
+    stack_size_record_t unopt_size_rec =
+        sm->stk_size_records[stmap_get_size_record(sm, unopt_rec_idx)];
+    // Locations are considered in pairs,  which is why the loop counter is
+    // incremented by 2. This is because locations stored at odd indices
+    // represent the sizes of the other locations. This is important because
+    // the runtime needs to know how many bytes to copy from each location to
+    // restore the stack state.
+    for (int i = 0; i < unopt_rec.num_locations - 1; i += 2) {
+        location_type type = unopt_rec.locations[i].kind;
+        // Get the value of the current location. This value needs to be placed
+        // on the stack of the unoptimized function, or in a register.
+        uint64_t opt_location_value =
+            stmap_get_location_value(sm, opt_rec.locations[i], r, bp);
+        if (type == REGISTER) {
+            uint16_t reg_num = unopt_rec.locations[i].dwarf_reg_num;
+            uint64_t loc_size =
+                stmap_get_location_value(sm, opt_rec.locations[i + 1], r, bp);
+            memcpy(r + reg_num, &opt_location_value, loc_size);
+        } else if (type == DIRECT) {
+            uint64_t loc_size =
+                stmap_get_location_value(sm, opt_rec.locations[i + 1], r, bp);
+            memcpy(direct_locations + i, &opt_location_value,
+                    loc_size);
+        } else if (type == INDIRECT) {
+            uint64_t unopt_addr = (uint64_t) bp + unopt_rec.locations[i].offset;
+            errx(1, "Not implemented - indirect.\n");
+        } else if (type != CONSTANT && type != CONST_INDEX) {
+            errx(1, "Unknown record - %u. Exiting\n", type);
+        }
+    }
+
+    // Populate the stack of the optimized function with the values the
+    // unoptimized function expects.
+    for (int i = 0; i < unopt_rec.num_locations - 1; i += 2) {
+        location_type type = unopt_rec.locations[i].kind;
+        if (type == DIRECT) {
+            uint64_t unopt_addr = (uint64_t)bp + unopt_rec.locations[i].offset;
+            uint64_t loc_size =
+                stmap_get_location_value(sm, opt_rec.locations[i + 1], r, bp);
+            memcpy((void *)unopt_addr, direct_locations + i, loc_size);
+        }
```

Comment:
Similarly here, what happens to other `type`s?

Reply:
The only `types` that are left are `CONST` and `CONST_INDEX`. If a value is a constant in the unoptimized trace, there is no need to restore it (since it's a constant, it should be there).